File: 1325-path-with-maximum-probability/path-with-maximum-probability.java
public class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start, int end) {
        List<List<Pair>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < edges.length; i++) {
            int a = edges[i][0];
            int b = edges[i][1];
            double prob = succProb[i];
            graph.get(a).add(new Pair(b, prob));
            graph.get(b).add(new Pair(a, prob));
        }

        double[] probabilities = new double[n];
        probabilities[start] = 1.0;

        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> Double.compare(b.probability, a.probability));
        pq.offer(new Pair(start, 1.0));

        while (!pq.isEmpty()) {
            Pair current = pq.poll();
            int node = current.node;
            double prob = current.probability;

            if (node == end) {
                return prob;
            }

            for (Pair neighbor : graph.get(node)) {
                int nextNode = neighbor.node;
                double edgeProb = neighbor.probability;
                double newProb = prob * edgeProb;

                if (newProb > probabilities[nextNode]) {
                    probabilities[nextNode] = newProb;
                    pq.offer(new Pair(nextNode, newProb));
                }
            }
        }

        return 0.0;
    }

    static class Pair {
        int node;
        double probability;

        Pair(int node, double probability) {
            this.node = node;
            this.probability = probability;
        }
    }
}

File: 1325-path-with-maximum-probability/path-with-maximum-probability.py
class Solution(object):
    def maxProbability(self, n, edges, succProb, start_node, end_node):
        graph = {}
        for i in range(n):
            graph[i] = []
        for i, edge in enumerate(edges):
            a, b = edge
            p = succProb[i]
            graph[a].append((b, p))
            graph[b].append((a, p))

        weights = [0] * n
        weights[start_node] = 1
        pq = []
        heapq.heappush(pq, (-1, start_node))

        while pq:
            weight, node = heapq.heappop(pq)
            weight = -weight
            if node == end_node:
                return weight
            for neighbor, prob in graph[node]:
                new_weight = weight * prob
                if new_weight > weights[neighbor]:
                    weights[neighbor] = new_weight
                    heapq.heappush(pq, (-new_weight, neighbor))

        return 0.0
